File: bench/corpus/sample.go
package main

import (
	"errors"
	"fmt"
	"sort"
	"strings"
	"sync"
	"time"
)

const maxRetries = 5

type Role string

const (
	RoleAdmin  Role = "admin"
	RoleMember Role = "member"
	RoleGuest  Role = "guest"
)

type User struct {
	ID        uint64
	Email     string
	Role      Role
	CreatedAt time.Time
}

func (u User) Describe() string {
	return fmt.Sprintf("%s <%s>", u.Email, u.Role)
}

type UserStore struct {
	mu    sync.RWMutex
	users map[uint64]User
	seq   uint64
}

func NewUserStore() *UserStore {
	return &UserStore{users: make(map[uint64]User)}
}

func (s *UserStore) Add(email string, role Role) User {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.seq++
	u := User{ID: s.seq, Email: strings.ToLower(email), Role: role, CreatedAt: time.Now()}
	s.users[u.ID] = u
	return u
}

func (s *UserStore) Find(id uint64) (User, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	u, ok := s.users[id]
	if !ok {
		return User{}, errors.New("not found")
	}
	return u, nil
}

func (s *UserStore) Filter(pred func(User) bool) []User {
	s.mu.RLock()
	defer s.mu.RUnlock()
	var out []User
	for _, u := range s.users {
		if pred(u) {
			out = append(out, u)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ID < out[j].ID })
	return out
}

func main() {
	store := NewUserStore()
	store.Add("a@example.com", RoleAdmin)
	store.Add("b@example.com", RoleMember)

	admins := store.Filter(func(u User) bool { return u.Role == RoleAdmin })
	for _, u := range admins {
		fmt.Println(u.Describe())
	}

	if u, err := store.Find(1); err == nil {
		fmt.Printf("found: %+v\n", u)
	} else {
		fmt.Println("error:", err)
	}
}

File: bench/corpus/sample.rs
use std::collections::HashMap;
use std::fmt;

const MAX_RETRIES: u32 = 5;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Role {
    Admin,
    Member,
    Guest,
}

impl fmt::Display for Role {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Role::Admin => "admin",
            Role::Member => "member",
            Role::Guest => "guest",
        };
        write!(f, "{s}")
    }
}

#[derive(Debug, Clone)]
pub struct User {
    pub id: u64,
    pub email: String,
    pub role: Role,
}

impl User {
    pub fn new(id: u64, email: impl Into<String>, role: Role) -> Self {
        Self {
            id,
            email: email.into(),
            role,
        }
    }

    pub fn describe(&self) -> String {
        format!("{} <{}>", self.email, self.role)
    }
}

#[derive(Default)]
pub struct UserStore {
    users: HashMap<u64, User>,
    seq: u64,
}

impl UserStore {
    pub fn add(&mut self, email: &str, role: Role) -> &User {
        self.seq += 1;
        let user = User::new(self.seq, email.to_lowercase(), role);
        self.users.entry(self.seq).or_insert(user)
    }

    pub fn find(&self, id: u64) -> Option<&User> {
        self.users.get(&id)
    }

    pub fn filter<F>(&self, pred: F) -> Vec<&User>
    where
        F: Fn(&&User) -> bool,
    {
        self.users.values().filter(pred).collect()
    }

    pub fn len(&self) -> usize {
        self.users.len()
    }
}

fn summarize<'a>(users: impl Iterator<Item = &'a User>) -> HashMap<String, usize> {
    let mut counts = HashMap::new();
    for user in users {
        *counts.entry(user.role.to_string()).or_insert(0) += 1;
    }
    counts
}

fn main() {
    let mut store = UserStore::default();
    store.add("a@example.com", Role::Admin);
    store.add("b@example.com", Role::Member);

    let admins: Vec<_> = store
        .filter(|u| u.role == Role::Admin)
        .iter()
        .map(|u| u.describe())
        .collect();

    println!("admins: {admins:?}");
    println!("summary: {:?}", summarize(store.filter(|_| true).into_iter().copied().collect::<Vec<_>>().iter().map(|u| u)));
    println!("total: {}", store.len());
}
